include/cm/nccl_ofi_cm.h

```diff
@@ -0,0 +1,228 @@
+	 */
+	nccl_ofi_connection_manager(fi_info *info, fid_domain *_domain,
+				    fid_cq *cq, size_t num_comm_ids,
+				    nccl_ofi_idpool_t *_mr_key_pool);
+
+	/**
+	 * Destructor. Finalizes CM endpoint and other state.
+	 *
+	 * Note: the connection manager should not be destroyed until all
+	 * associated cm communicators (TODO: rename from communicators) are
+	 * destroyed
+	 */
+	~nccl_ofi_connection_manager();
+
+	/**
+	 * Return a communicator (TODO rename to handle or some such) that can
+	 * be used to accept incoming connections from connect()
+	 */
+	nccl_ofi_cm_l_comm *listen();
+
+	/**
+	 * Given a handle (produced by cm_l_comm on a remote node), establish
+	 * a (cm_s_comm) connection to the remote node. This will return a
+	 * cm_s_comm handle that can be queried until the connection is fully
+	 * established.
+	 *
+	 * @param handle: handle from cm_l_comm on a remote node
+	 * @param rail_info: rail information (addresses of data-transfer endpoints to send
+	 * 		     to the remote node)
+	 */
+	nccl_ofi_cm_s_comm *connect(nccl_net_ofi_conn_handle *handle,
+				    const nccl_ofi_cm_ep_rail_info &rail_info);
+
+
+	/**
+	 * Provide a "rail selector" used to choose rails to advertise
+	 * in the connect response message.
+	 */
+	void set_rail_selector(select_rails_fn fn, void *fn_input)
+	{
+		rail_selector = fn;
+		rail_selector_input = fn_input;
+	}
+	/* --------------------------------------------------------- */
+
+	/* TODO: the functions below are not intended to be used by the caller
+	   (i.e., transport), but are public due to use by other classes in the
+	   cm code. Revisit the separation of concerns here. */
+
+	/**
+	 * Map from comm id to l_comm. Used for receiving connect message
+	 */
+	nccl_ofi_cm_l_comm *get_l_comm(uint32_t l_comm_id);
+	/**
+	 * Map from comm id to s_comm. Used for receiving connect response message
+	 */
+	nccl_ofi_cm_s_comm *get_s_comm(uint32_t s_comm_id);
+
+	/**
+	 * Access functions for the conn_msg freelist (which takes care of memory
+	 * registration)
+	 */
+	nccl_ofi_freelist_elem_t *alloc_conn_msg()
+	{
+		return nccl_ofi_freelist_entry_alloc(conn_msg_fl.get());
+	}
+	void free_conn_msg(nccl_ofi_freelist_elem_t *conn_msg)
+	{
+		nccl_ofi_freelist_entry_free(conn_msg_fl.get(), conn_msg);
+	}
+
+	/**
+	 * Accessors for OFI resources
+	 */
+	fid_ep *get_ep() {return ep;}
+	fid_domain *get_domain() {return domain;}
+
+	/**
+	 * Accessor for CM-managed address vector. Used by cm_s_comm and cm_l_comm
+	 * to insert address from received connect/connect-response messages
+	 *
+	 * TODO just make an accessor for fid_av* instead?
+	 */
+	int av_insert_address(const ep_name address, fi_addr_t *fi_addr);
+
+	const cm_ep_name &get_conn_ep_name() {return conn_ep_name;}
+
+	nccl_ofi_idpool_t *get_l_comm_id_pool() { return &l_comm_id_pool; }
+	nccl_ofi_idpool_t *get_data_comm_id_pool() { return &data_comm_id_pool; }
+
+	std::unordered_map<uint32_t, nccl_ofi_cm_l_comm *> *get_l_comm_map()
```

Comment:
Let's say tomorrow you decide that you need the connection lookup to be linear instead of logarithmic.  So you change the type of the conn map to be a vector.  Now you've got to change this interface function as well, which just makes no sense from an abstraction point of view.